Add the ability to simply extract the ToC in Markdown format.

TomOnTime opened the pull request:
This is related to https://github.com/russross/blackfriday/issues/201

rtfb commented:
Great, let's wait what @shurcooL has to say about it :-)

shurcooL commented:
I've looked over this and the proposed change to `sanitized_anchor_name`.

I don't think that changing `sanitized_anchor_name` is the best way to solve this task. The update is specific to how GitLab happens to perform their header title -> sanitized anchor name mapping. If someone else later wants to make their anchors compatible with another style associated with a different company, they'd have to change `sanitized_anchor_name` yet again and it'd break the new GitLabs-flavored behavior.

Since what's being added here is a new optional feature for specific users to opt into, I think the best way is to add a new field to [`Options`](https://godoc.org/github.com/russross/blackfriday#Options) type:

```Go
// Options represents configurable overrides and callbacks (in addition to the
// extension flag set) for configuring a Markdown parse.
type Options struct {
	...

	// SanitizedAnchorNameOverride is an optional func to override
	// the behavior for creating sanitized anchor names.
	SanitizedAnchorNameOverride SanitizedAnchorNameFunc
}

// SanitizedAnchorNameFunc creates sanitized anchor names.
type SanitizedAnchorNameFunc func(text string) string
```

By default (if `SanitizedAnchorNameOverride` is left as `nil`), the current `sanitized_anchor_name.Create` can be used to provide a simple default behavior.

If GitLab-specific behavior for creating anchor names is desired, that func can be set by the user to override and get the behavior they want.

What do you think of that approach?

The main reason I don't want to change `sanitized_anchor_name` is because the changes are not specific to a widely accepted standard but rather an arbitrary set of choices that are specific to GitLab only, as far as I can tell. Given that, I expect there would be similar variations elsewhere. `sanitized_anchor_name` is mean to be a very simple default. But if you disagree here, please let me know.

TomOnTime commented:
Sounds good!  Though I haven't tested GitHub or other systems to see if they use the same algorithm.

TomOnTime commented:
Ok, I did some tests with Github and it does have differences:

https://github.com/TomOnTime/tomontime/wiki/test_headers
https://gitlab.com/tal/gitlab-markdown-test/wikis/gitlab_headers

Here are the cases where things are generated differently:
https://gist.github.com/TomOnTime/eff1439aa0682f34653b

How about I add CreateGitLab() and CreateGitHub()  to sanitized_anchor_name/main.go and make main_test.go have both test suites?

TomOnTime commented:
Assuming the changes to https://github.com/shurcooL/sanitized_anchor_name are accepted, this should now let you select the anchor sanitizer of your choosing.

shurcooL commented on markdown.go:
```diff
@@ -374,6 +387,12 @@ func MarkdownOptions(input []byte, renderer Renderer, opts Options) []byte {
 	}
 
+	if opts.SanitizedAnchorNameOverride == nil {
+		p.sanitizedAnchorNameOverride = sanitized_anchor_name.Create
```
This logic could be done in one place (in html.go) instead of two.

shurcooL commented on html.go:
```diff
@@ -106,12 +114,12 @@ const (
 
 func HtmlRendererWithParameters(flags int, title string,
-	css string, renderParameters HtmlRendererParameters) Renderer {
+	css string, sanitize SanitizedAnchorNameFunc, renderParameters HtmlRendererParameters) Renderer {
```
Why did you choose to change the signature of `HtmlRendererWithParameters` instead of adding the `SanitizedAnchorNameFunc` as a new field to `HtmlRendererParameters`?

TomOnTime commented:
> Why did you choose to change the signature of HtmlRendererWithParameters instead of adding the
> SanitizedAnchorNameFunc as a new field to HtmlRendererParameters?

Short answer: I wanted to but I didn't see how to do it and still get the info from blackfriday-tool to that place.  Hints?

rtfb commented:
BTW, it suddenly struck me that this functionality fits ideally in https://github.com/shurcooL/markdownfmt, while in Blackfriday it introduces a completely new form of output. Also, you wouldn't even need a new flag in markdownfmt, only make it respect `HTML_OMIT_CONTENTS`.

TomOnTime commented:
@rtfb 

I've push a change that addresses the concerns (HtmlRendererWithParameters no longer has a new signature, etc).  Sadly it breaks one test in a way that I can't figure out how to fix.  Do you have any suggestions?

As far as adding the functionality to https://github.com/shurcooL/markdownfmt instead, I'd be interested in @shurcooL's opinion.  It seems like that would require a lot of code changes.

@shurcooL I see two ways to make this work as part of markdownfmt instead.  One would be to simply have HTML_OMIT_CONTENTS omit most output, but when outputting Headers instead output the ToC entry.  The other would be to look for a token pair such as [TOC][/TOC] and replace the contents with the ToC.  The latter seems like a better way to do it, but a lot more work (and beyond my understanding of how to add that function).  Thoughts?

shurcooL commented:
About https://github.com/shurcooL/markdownfmt, the `markdownfmt` command is designed not to have any formatting options (just like `gofmt` doesn't have formatting options), so I don't think it would work. It's also outside the scope of that project. You're welcome to fork it and modify it to be a dedicated tool for what you're doing, if you want.

---

> Assuming the changes to https://github.com/shurcooL/sanitized_anchor_name are accepted, this should now let you select the anchor sanitizer of your choosing.

I've reviewed the latest changes at https://github.com/shurcooL/sanitized_anchor_name/pull/2#issuecomment-149087410, please see that comment.

As a followup, unless I'm mistaken, no changes to the current `sanitized_anchor_name` package need to be accepted for this PR to be able to go through. Additional sanitize-anchor-name functionality can be in any Go package, not necessarily that one. Is that correct? In https://github.com/shurcooL/sanitized_anchor_name/pull/2#issuecomment-149087410, I've outlined the reasons why I don't think it's the best home for it.

rtfb commented:
So what's the status of this, @TomOnTime? Are you satisfied with how things ended up partitioned? This PR looks good now, I can merge if you think it's what it should be.

shurcooL commented on block_test.go:
```diff
@@ -35,6 +37,10 @@ func runnerWithRendererParameters(parameters HtmlRendererParameters) func(string
+			parameters.SanitizedAnchorNameOverride = sanitized_anchor_name.Create
+		}
+
 		renderer := HtmlRendererWithParameters(htmlFlags, "", "", parameters)
```
Are the 4 added lines above needed? `HtmlRendererWithParameters` already sets the default value for `SanitizedAnchorNameOverride` if it's `nil`.

TomOnTime commented:
I haven't had a chance to look at this.  I'll get to it in the next few days.

TomOnTime commented:
I made the changes as requested. Also, the GitLab and GitHub functionality was split into markdownutils, so I've adjusted the code as required.  This is ready for merge.

shurcooL commented on html.go:
```diff
@@ -70,6 +74,9 @@ type HtmlRendererParameters struct {
 	HeaderIDSuffix string
+	// SanitizedAnchorNameOverride is an optional func to override
+	// the behavior for creating sanitized anchor names.
+	SanitizedAnchorNameOverride SanitizedAnchorNameFunc
```
This PR should be about adding the _option_ to change the default. These 3 lines are good.

shurcooL commented on html.go:
```diff
@@ -34,6 +37,7 @@ const (
 	HTML_NOREFERRER_LINKS                      // only link with rel="noreferrer"
 	HTML_HREF_TARGET_BLANK                     // add a blank target
 	HTML_TOC                                   // generate a table of contents
+	HTML_TOC_MD                                // if generating table of contents, output MarkDown
```
This library spells Markdown as "Markdown", not "MarkDown". This should be kept consistent.

It's not very clear what this flag does. Output Markdown instead of what?

TomOnTime replied:
Fixed.

shurcooL commented on html.go:
```diff
@@ -123,6 +130,10 @@ func HtmlRendererWithParameters(flags int, title string,
 	}
 
+	if renderParameters.SanitizedAnchorNameOverride == nil {
+		renderParameters.SanitizedAnchorNameOverride = markdownutils.CreateGitLabAnchor
```
What's the motivation to change the default? This seems unnecessary. Users will be able to set it to `markdownutils.CreateGitLabAnchor` or another custom func by using the `SanitizedAnchorNameOverride` parameter.

TomOnTime replied:
Removed.  My intention was to avoid a nil reference, but testing the code with those 3 lines removed works so I guess it wasn't needed.

shurcooL replied:
Perhaps you misunderstood. I meant that `blackfriday` should continue to use `sanitized_anchor_name.Create` as the default anchor sanitizer. This PR should not change the default to `markdownutils.CreateGitLabAnchor`, that should be done in your own code that imports and uses `blackfriday`.

shurcooL commented:
> * Use github.com/TomOnTime/markdownutils instead of sanitized_anchor_name

I don't think this PR should be making that change.

TomOnTime commented:
Ah!  Now I get it.  The dependency isn't needed, so it can removed.  (Now I understand what you were saying before)

shurcooL commented on block.go:
```diff
@@ -243,7 +243,11 @@ func (p *parser) prefixHeader(out *bytes.Buffer, data []byte) int {
+				id = sanitized_anchor_name.Create(string(data[i:end]))
+			} else {
+				id = p.sanitizedAnchorNameOverride(string(data[i:end]))
+			}
```
I think it'd be cleaner to ensure `p.sanitizedAnchorNameOverride` is never nil (by giving it a default value when there's no override), so the code here can be just:

```Go
id = p.sanitizedAnchorNameOverride(string(data[i:end]))
```

At that point, you'd want to rename `sanitizedAnchorNameOverride` to just `sanitizedAnchorName`.

If an override is provided, its value will be set to:

```Go
sanitizedAnchorName = sanitized_anchor_name.Create
```

shurcooL commented on block.go:
```diff
@@ -1346,7 +1350,11 @@ func (p *parser) paragraph(out *bytes.Buffer, data []byte) int {
+						id = sanitized_anchor_name.Create(string(data[prev:eol]))
+					} else {
+						id = p.sanitizedAnchorNameOverride(string(data[prev:eol]))
+					}
```
Same here, this could become just:

```Go
id = p.sanitizedAnchorName(string(data[i:end]))
```

shurcooL commented on html.go:
```diff
@@ -483,9 +488,15 @@ func (options *Html) AutoLink(out *bytes.Buffer, link []byte, kind int) {
+		out.WriteString("`")
+		out.Write(text)
+		out.WriteString("`")
+	}
```
I don't like the extra complexity this brings, but I don't see a different way of doing this if we're to support this feature.

shurcooL commented on markdown.go:
```diff
@@ -380,6 +392,8 @@ func MarkdownOptions(input []byte, renderer Renderer, opts Options) []byte {
 		p.notes = make([]*reference, 0)
 	}
 
+	p.sanitizedAnchorNameOverride = opts.SanitizedAnchorNameOverride
```
This would be where you'd check if `opts.SanitizedAnchorNameOverride` is `nil` and use an internal default value if so:

```Go
if opts.SanitizedAnchorNameOverride != nil {
    p.sanitizedAnchorName = opts.SanitizedAnchorNameOverride
} else {
    p.sanitizedAnchorName = sanitized_anchor_name.Create
}
```

shurcooL commented:
> Ah! Now I get it. The dependency isn't needed, so it can removed. (Now I understand what you were saying before)

Great! The latest version is starting to look a lot better! I've left a few more suggestions, what do you think about them?

TomOnTime commented:
I'll try to get to it this week.

jonny-improbable commented:
Did anything ever come of this feature?

TomOnTime commented:
Sorry, I'm going to have to decline to continue working on this.  If someone else is interested, I'd be glad to help.  Sadly the project I was going to use this for is no longer happening and I won't have time to finish the PR.

shurcooL commented:
Thanks for letting us know @TomOnTime, and thank you for the work to get it to this point!
